src/TuningSystem.cpp

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright © 2014-2023 Synthstrom Audible Limited
+ *
+ * This file is part of The Synthstrom Audible Deluge Firmware.
+ *
+ * The Synthstrom Audible Deluge Firmware is free software: you can redistribute it and/or modify it under the
+ * terms of the GNU General Public License as published by the Free Software Foundation,
+ * either version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program.
+ * If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include <math.h>
+#include "TuningSystem.h"
+
+int32_t tuningFrequencyTable[12];
+int32_t tuningIntervalTable[12];
+
+TuningSystem tuningSystem;
+
+int32_t selectedTuningBank;
+
+const double sampleRateDiv4 = 11025.0;
+
+void TuningSystem::calculateOffset(int noteWithinOctave) {
+
+	double cents = 100 * (noteWithinOctave - 5);
+	cents += offsets[noteWithinOctave] / 100.0;
+	double frequency = referenceFrequency * pow(2.0, cents / 1200.0);
+
+
+	double value = frequency / 1378.125;
+	value *= pow(2.0, 32);
+	tuningFrequencyTable[noteWithinOctave] = lround(value);
+
+	value = pow(2.0, noteWithinOctave / 12.0) * pow(2.0, 30);
+	tuningIntervalTable[noteWithinOctave] = lround(value);
```

Comment:
This should be readily adaptable to precise tuning across the entire range, and this function is only invoked by turning the knob so no performance hit expected. I've verified (on x86_64 at least!) that when offsets are all zero, the same values are generated as in lookuptables.cpp. I'd like to add unit tests for things like this, but my current test source isn't pretty enough to work on everyone's machine, and it'd be better if I were compiling for the same target and running the test binary on QEMU.